transport/tchannel/handler.go

```diff
@@ -167,101 +174,128 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, start time.T
+		err = transport.DispatchUnaryHandler(ctx, spec.Unary(), start, treq, responseWriter)
 
 	default:
-		err = errors.UnsupportedTypeError{Transport: "TChannel", Type: spec.Type().String()}
+		err = yarpcerrors.UnimplementedErrorf("transport:tchannel type:%s", spec.Type().String())
 	}
 
 	return err
 }
 
 type responseWriter struct {
-	treq         *transport.Request
-	failedWith   error
-	bodyWriter   tchannel.ArgWriter
-	format       tchannel.Format
-	headers      transport.Headers
-	response     inboundCallResponse
-	wroteHeaders bool
+	// TODO: do we still need this?
+	failedWith error
+	format     tchannel.Format
+	headers    transport.Headers
+	buffer     *bytes.Buffer
+	response   inboundCallResponse
 }
 
-func newResponseWriter(treq *transport.Request, call inboundCall) *responseWriter {
+func newResponseWriter(response inboundCallResponse, format tchannel.Format) *responseWriter {
 	return &responseWriter{
-		treq:     treq,
-		response: call.Response(),
-		format:   call.Format(),
+		response: response,
+		format:   format,
 	}
 }
 
 func (rw *responseWriter) AddHeaders(h transport.Headers) {
-	if rw.wroteHeaders {
-		panic("AddHeaders() cannot be called after calling Write().")
-	}
 	for k, v := range h.Items() {
-		rw.headers = rw.headers.With(k, v)
+		// TODO: is this considered a breaking change?
+		if isReservedHeaderKey(k) {
+			panic("cannot use reserved header key " + k)
+		}
+		rw.addHeader(k, v)
 	}
 }
 
+func (rw *responseWriter) addHeader(key string, value string) {
+	rw.headers = rw.headers.With(key, value)
+}
+
 func (rw *responseWriter) SetApplicationError() {
-	if rw.wroteHeaders {
-		panic("SetApplicationError() cannot be called after calling Write().")
-	}
 	err := rw.response.SetApplicationError()
 	if err != nil {
+		// TODO: just set failedWith?
 		panic(fmt.Sprintf("SetApplicationError() failed: %v", err))
 	}
 }
 
-func (rw *responseWriter) ensureWroteHeaders() error {
-	if rw.wroteHeaders {
-		return nil
-	}
-
-	rw.wroteHeaders = true
-	if err := writeHeaders(rw.format, rw.headers, rw.response.Arg2Writer); err != nil {
-		err = encoding.ResponseHeadersEncodeError(rw.treq, err)
-		rw.failedWith = err
-		return err
-	}
-
-	return nil
-}
-
 func (rw *responseWriter) Write(s []byte) (int, error) {
 	if rw.failedWith != nil {
 		return 0, rw.failedWith
 	}
 
-	if err := rw.ensureWroteHeaders(); err != nil {
-		return 0, err
+	if rw.buffer == nil {
+		rw.buffer = buffer.Get()
 	}
 
-	if rw.bodyWriter == nil {
-		var err error
-		rw.bodyWriter, err = rw.response.Arg3Writer()
-		if err != nil {
-			rw.failedWith = err
-			return 0, err
-		}
-	}
-
-	n, err := rw.bodyWriter.Write(s)
+	n, err := rw.buffer.Write(s)
 	if err != nil {
 		rw.failedWith = err
 	}
 	return n, err
 }
 
-func (rw *responseWriter) Close() error {
-	err := rw.ensureWroteHeaders()
-
-	if rw.bodyWriter != nil {
-		err = multierr.Append(err, rw.bodyWriter.Close())
+func (rw *responseWriter) Close(hasSystemError bool) error {
```

Comment:
We shouldn't have a boolean parameter for closing, if we need two different types of closing we should have two functions, or we can just have this function take the handler error directly.